Element Ingest

garvitverma opened the pull request:
Element Ingest.

mbrainerd commented on hooks/tk-multi-publish2/ingest/collector.py:
```diff
@@ -0,0 +1,257 @@
+import sgtk
+from sgtk import TankError
+
+from tank import context
```
Remove this import

mbrainerd commented on hooks/tk-multi-publish2/ingest/collector.py:
```diff
@@ -0,0 +1,257 @@
+            properties["work_path_template"] = work_path_template
+
+        # build the context of the item
+        context = self._get_item_context_from_path(parent_item, properties, os.path.basename(path))
```
Rather than overriding _add_file_item(), why don't you just override _get_item_context_from_path() and _resolve_work_path_template() to run the os.path.basename() and then call the parent superclass?

garvitverma replied:
Makes sense, making the change.

mbrainerd commented on hooks/tk-multi-publish2/ingest/collector.py:
```diff
@@ -0,0 +1,257 @@
+
+
+    def _resolve_item_fields(self, item):
+        """
```
Why is this method overridden?

mbrainerd replied:
I did you one better and broke out the template parsing part into a new overwrite-able sub-method `_get_template_fields_from_path()`, so you can just do something similar to what you did for `_resolve_work_path_template()`. Its available in `sgtk_apps-2.7.2`

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+import sgtk
+from sgtk import TankError
+
+from tank.util.shotgun import get_sg_connection
```
Please use the api method instead and remove this import.

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+
+    @staticmethod
+    def _validate_plate_entity(item):
+        sg = get_sg_connection()
```
Please use the api method instead: self.sgtk.shotgun.find_one()

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+        return first_frame, last_frame
+
+    def _create_plate_entity(self, item):
+        sg = get_sg_connection()
```
Please use the api method instead: self.sgtk.shotgun.find_one()

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+        :param item:
+        :return:
+        """
+        sg = get_sg_connection()
```
Please use the api method instead: self.sgtk.shotgun.find_one()

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+        contain simple html for formatting.
+        """
+
+        desc = super(IngestFilesPlugin, self).description
```
The parent description is describing a publish process, which (even though this class inherits from) is not an accurate description for what is happening here. Since this is user-facing, please make sure you fully describe (so that an artist or producer can understand) exactly what this process is doing.

garvitverma replied:
Updated this.

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+
+        super(IngestFilesPlugin, self).publish(task_settings, item)
+
+        # create a plate entity after the publish has went through successfully.
```
"has gone through". Also, you should probably test that the parent process has succeeded before continuing on.

garvitverma replied:
you want me to check for "sg_publish_data" on the item? Or is there any better way to know if the process succeeded?

mbrainerd replied:
That's the best way.

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+        ]
+        if item.context.entity:
+            if item.context.entity["type"] == "Shot":
+                sg_filters.append(['shots', 'is', item.context.entity])
```
Use `sg_shot` not `shots`

garvitverma replied:
Any reason we are not updating this field? We were updating this in plate_ingest.

mbrainerd replied:
`shots` should/will be used for specifying instances of that plate, not the plate entity itself

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+        )
+
+    @staticmethod
+    def _validate_plate_entity(item):
```
Missing docstring. Also, why is this called "validate"? There is no validation happening in this method. Perhaps "get" or "find"?

garvitverma replied:
Added docstring, changed method name to "find"

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+            filters=sg_filters,
+            fields=['shots', 'code', 'id']
+        )
+        return result
```
Perhaps throw an exception if the result is Null?

garvitverma replied:
This method is being used in _create_plate_entity, if it doesn't find a plate_entity corresponding to the item. It will create a new entity.

mbrainerd replied:
The SG python api will at times throw an exception. Wrapping the call to _find_plate_entity() in a try/except, adding your own exception if the result is null, and then handling the exception in _create_plate_entity() is just a more robust solution.

garvitverma replied:
If it throws an Exception, should we still proceed by just creating a new plate? Or should we stop the ingestion completely, and raise an error?

Does that also mean all the SG python api can throw exceptions?

mbrainerd replied:
I would think stop the ingestion for that item and log an error, but don't raise an exception so that the rest can proceed, but I leave it up to you as to how critical you think it should be.

mbrainerd replied:
Yes, all python api calls can throw exceptions.

garvitverma replied:
I can log it as an error and let the other items continue. But, what if the publish entity gets created but the plate entity fails. Do we need to change the validation/publish as well? So that it works with an existing publish file and linking it to the plate? Like the plate_ingest does too.

mbrainerd replied:
I would actually suggest switching the order so the publishedfile creation occurs after the plate entity creation, so that you can handle cleaning up the plate entity if the publishedfile entity creation fails.

garvitverma replied:
Makes sense, there will be only one case left then if the publish file and plate entity both get created... but the linking fails. But we can't re-run it just like that since the duplicate publish file on disk validation will fail.

And it won't show up in the conflicting publish list either as the finalize will run for that item and clear the status, since we are handling the exception. Thoughts?

mbrainerd replied:
Not sure I fully follow, but ultimately, since you control the calling method, you should have access to all involved entities, as well as their linked files on disk, so if anything fails you should be able to clean it all up before exiting the process. Quite frankly, the parent class should probably be handling its own cleanup, if its not already...can't remember, in the case of failure.

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+
+        return first_frame, last_frame
+
+    def _create_plate_entity(self, item):
```
Missing docstring

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+        )
+        return result
+
+    def _get_frame_range(self, item):
```
Missing docstring

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+            sg_client_name=item.name,
+        )
+
+        data["cut_in"] = frange[0]
```
You sure this should be cut_in/cut_out and not head_in/head_out?

garvitverma replied:
Yeah on the Element entity, those are the fields that the loader uses when creating the DCC nodes.

mbrainerd replied:
Let's discuss this at our meeting today.

garvitverma replied:
Switching this to update head_in and head_out fields now, we will need to update loader to be giving priority to these keys from now on, will open a ticket for that too.

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+
+        if item.context.entity:
+            if item.context.entity["type"] == "Shot":
+                data["shots"] = [item.context.entity]
```
Do not populate this field

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+
+            data["project"] = item.context.project
+            plate_entity = sg.create(entity_type='Element', data=data)
+            self.logger.info("Created Plate entity...")
```
No need for both info and debug. I would just use info

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+                data=data,
+                multi_entity_update_modes=dict(shots='add'),
+            )
+            self.logger.info("Updated Plate entity...")
```
No need for both info and debug. I would just use info

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+        super(IngestFilesPlugin, self).publish(task_settings, item)
+
+        # create a plate entity after the publish has went through successfully.
+        plate_entity = self._create_plate_entity(item)
```
Need exception handling for invalid plate entity

garvitverma replied:
If the plate entity doesn't exists, it creates it. If it finds it, it will get updated... I'm not sure how we will get an invalid plate.

mbrainerd replied:
The Shotgun api can (and does) throw the odd exception now and again.

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+        item.properties["ingest_plate_data"] = plate_entity
+
+        # link the publish file to our plate entity.
+        self._link_published_files_to_plate_entity(plate_entity, item)
```
Needs exception handling.

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+
+import os
+import copy
+import glob
```
unused import

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import copy
```
unused import

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,198 @@
+import copy
+import glob
+import pprint
+import traceback
```
unused import

mbrainerd commented on hooks/tk-multi-publish2/ingest/collector.py:
```diff
@@ -0,0 +1,169 @@
+         in this class we use os.path.basename of the path.
+        :return: Name of the template.
+        """
+
```
Remove extra whitespace

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,220 @@
+        """
+
+        return """
+        Ingests the file to the location specified by the templates and
```
"...specified by the publish_path_template defined for this item and..."

garvitverma replied:
Isn't that producer or artist facing documentation? publish_path_template is inside the templates which we are going to setup anyways.

mbrainerd replied:
Fair enough, though putting publish_path_template is useful for a TD trying to debug :)

garvitverma replied:
Hehe ok, adding it for the TD team :)

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,220 @@
+        creates a <b>PublishedFile</b> entity in Shotgun, which will include a
+        reference to the file's published path on disk.
+
+        After the <b>PublishFile</b> is created successfully, an <b>Element/Plate</b> entity is also created.
```
Just call it a Plate entity, having both is confusing.

mbrainerd replied:
Also, PublishedFile, not PublishFile

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,220 @@
+        reference to the file's published path on disk.
+
+        After the <b>PublishFile</b> is created successfully, an <b>Element/Plate</b> entity is also created.
+        The PublishFile is then linked to it's corresponding Element entity for other users to use.
```
PublishedFile, not PublishFile, and Plate entity, not Element entity

mbrainerd commented on hooks/tk-multi-publish2/ingest/ingest_files.py:
```diff
@@ -0,0 +1,220 @@
+
+        After the <b>PublishFile</b> is created successfully, an <b>Element/Plate</b> entity is also created.
+        The PublishFile is then linked to it's corresponding Element entity for other users to use.
+        Other users will be able to access the published file via the plates.
```
This sentence is too vague, please update.

garvitverma commented:
Pushed the config for element ingest to internal git on same branch.
Updated ingest_files.py to implement undo method for deleting plate entity that got created if the process fails.

garvitverma commented:
- Rebased the commits.... since I wasn't able to rebase them over the master.
- Pushed the config for element ingest to internal git on same branch.
Updated ingest_files.py to implement undo method for deleting plate entity that got created if the process fails.
